### PR TITLE
[posix] refactor posix platform modules

### PR DESCRIPTION
**Pull request by wgtdkp**
This commit refactors the posix platform implementation with below changes:

1. Remove use of global OT instance (i.e. `gInstance`). It's not safe to use `gInstance`
    in the posix platform because Platform code can run before the OT instance is
    created/initialized.
2. Change `XxxxSetUp()` methods to accept a `otInstance *aInstance` so that
    we know the OT instance is available to use only when `SetUp()` is called.
3. define classes for sub-modules that need to call OT public or platform APIs.
    This also follow the style in OT core.
4. Replace global / static variables with singleton class or member fields of the
    sub-module classes

Depends-On: openthread/ot-br-posix#1944

### CODE REVIEW

File: src/posix/platform/multicast_routing.cpp [hunk 1]
```diff
@@ -69,21 +69,33 @@ namespace Posix {
         }                                                                                                           \
     } while (false)
 
-void MulticastRoutingManager::SetUp(void)
+MulticastRoutingManager &MulticastRoutingManager::Get(void)
 {
-    OT_ASSERT(gInstance != nullptr);
+    static MulticastRoutingManager sManager;
```
[the remark this answers is not in the document]
Comment (wgtdkp) [reply src/posix/platform/multicast_routing.cpp:74]: Thanks for calling this out.
I think this implies that we shouldn't do work other than member initialization in constructors. Initialization code should be in the `Init()` or `SetUp()` method.

I was initializing the mainloop (i.e. `Mainloop::Manager::Get().Add(*this);`) for two modules in their constructors and now I have removed them out.

File: src/posix/platform/netif.hpp [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+namespace ot {
+namespace Posix {
+
+class NetIf : public Mainloop::Source, private NonCopyable
+{
+public:
+    /** Returns the singleton object of this class. */
+    static NetIf &Get(void);
+
+    void Init(const otPlatformConfig *aPlatformConfig);
```
Comment (jwhui) [inline src/posix/platform/netif.hpp:56]: Add documentation for public methods?

File: src/posix/platform/udp.hpp [hunk 3]
```diff
@@ -39,12 +41,13 @@ class Udp : public Mainloop::Source, private NonCopyable
 public:
     static Udp &Get(void);
 
-    void Init(const char *aIfName);
-    void SetUp(void);
+    void SetUp(otInstance *aInstance);
```
Comment (jwhui) [inline src/posix/platform/udp.hpp:44]: Add documentation for public methods?

File: src/posix/platform/daemon.hpp [hunk 4]
```diff
@@ -41,7 +41,7 @@ class Daemon : public Mainloop::Source, private NonCopyable
 public:
     static Daemon &Get(void);
 
-    void SetUp(void);
+    void SetUp(otInstance *aInstance);
```
Comment (jwhui) [inline src/posix/platform/daemon.hpp:44]: Add documentation for public methods?

File: src/posix/platform/mld_monitor.hpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+namespace ot {
+namespace Posix {
+
+class MldMonitor : public Mainloop::Source, private NonCopyable
+{
+public:
+    /** Returns the singleton object of this class. */
+    static MldMonitor &Get(void);
+
+    void SetUp(otInstance *aInstance);
```
Comment (jwhui) [inline src/posix/platform/mld_monitor.hpp:50]: Add documentation for public methods?

File: src/posix/platform/infra_if.hpp [hunk 6]
```diff
@@ -84,7 +77,7 @@ class InfraNetif : public Mainloop::Source, private NonCopyable
      * @note This method is called after OpenThread instance is created.
      *
      */
-    void SetUp(void);
+    void SetUp(otInstance *aInstance);
```
[the remark this answers is not in the document]
Comment (wgtdkp) [reply src/posix/platform/infra_if.hpp]: done

File: src/posix/platform/include/openthread/openthread-system.h [hunk 7]
```diff
@@ -182,31 +180,29 @@ extern otPlatResetReason gPlatResetReason;
...
- * @returns The Thread network interface name.
- *
  */
 const char *otSysGetThreadNetifName(void);
 
 /**
- * Returns the Thread network interface index.
- *
- * @returns The Thread network interface index.
+ * Returns the Thread network interface index..
```
[the remark this answers is not in the document]
Comment (wgtdkp) [reply src/posix/platform/include/openthread/openthread-system.h]: done

File: src/posix/platform/include/openthread/openthread-system.h [hunk 8]
```diff
@@ -172,8 +172,6 @@ void otSysMainloopProcess(otInstance *aInstance, const otSysMainloopContext *aMa
 /**
  * Returns the radio url help string.
  *
- * @returns the radio url help string.
```
[the remark this answers is not in the document]
Comment (wgtdkp) [reply src/posix/platform/include/openthread/openthread-system.h:175]: Those looks redundant for a simple getter API - it's simply a copy of the summary sentence.

Do you want doxygen commands like `@returns` always present?

File: src/posix/platform/include/openthread/openthread-system.h [hunk 9]
```diff
@@ -123,7 +123,7 @@ otInstance *otSysInit(otPlatformConfig *aPlatformConfig);
  *       when deinitialization of OpenThread's drivers is most appropriate.
  *
  */
-void otSysDeinit(void);
+void otSysDeinit(otInstance *aInstance);
```
[the remark this answers is not in the document]
Comment (wgtdkp) [reply src/posix/platform/include/openthread/openthread-system.h:128]: done

File: src/posix/platform/udp.hpp (hunk 3 above)
Comment (wgtdkp) [reply src/posix/platform/udp.hpp]: I changed to define `SetUp()` and `TearDown()` as base methods in the 3rd commit of this commit

File: src/posix/platform/netif.hpp (hunk 2 above)
Comment (wgtdkp) [reply src/posix/platform/netif.hpp:63]: done

**Pull request closed without merging**